_src/_posts/2018-09-07-writing-custom-presto-functions.markdown

```diff
@@ -0,0 +1,92 @@
+---
+layout: post
+title:  Overview of Jampp's platform
+date:   
+tag:    technology
+categories:
+    - infrastructure
+    - technology
+keywords: "Presto, PrestoDB, UDF, aggregation"
+author: dantepawlow
+---
+
+# Writing Custom PrestoDB Functions
+
+<!--excerpt.start-->
+Here at Jampp we process and analyse large amounts of data. One of the tools we employ to do so is [PrestoDB](https://prestodb.io/), which is a "Distributed SQL Query Engine for Big Data".
+Presto comes with many [functions](https://prestodb.io/docs/current/functions.html), which are usually enough for most use cases. Nevertheless, sometimes you need to implement your own function for a very specific use.
+
+Enter the [User Defined Functions](https://prestodb.io/docs/current/develop/functions.html) (UDFs, for short). Writing one is not as straightforward as it first appears, mainly because the information to do so is very scattered around the web (and across many Presto versions).
+
+In this blogpost, we present our _JSON Sum_ function, how did we write it, and some of the lessons we learned along the way.
+<!--excerpt.end-->
+
+## Function types
+
+Presto has two main types of functions: scalar and aggregation[^1]. Scalar functions behave more like traditional functions, in the sense that their output is immediate from the values of their parameters. On the other hand, aggregation functions take multiple rows as input and combine them into a single output, we'll focus mainly on these functions, as they're more complex (and, interesting!).
+
+[^1]: There's also window functions, but we have yet to implement one of those, so we'll leave them out of the scope of this article.
+
+Aggregation functions can harness the power of presto’s distributed workers via a divide and conquer approach.
+They consist of three main methods:
+
+    - _input_, which reads each row and feeds it to a state.
+    - _combine_, invoked whenever two states need to be merged.
+    - _output_, which returns the result of the aggregation at the end.
+
+The _state_ is a data structure that stores the meaningful results of each step of the aggregation, and can be designed according to the specific needs. It should be serializable (and deserializable) to share it between nodes.
+
+It's important to note that no order of input or combination should be assumed: any given state should be able to be combined with any other and any row should be able to be converted to a valid state. You can, however, define the initial parameters of the state via its constructor.
+
+## Implementation
+
+UDFs are implemented inside a [plugin](https://prestodb.io/docs/current/develop/spi-overview.html), separate from the default Presto functions. They are exposed to presto by declaring them in [UdfPlugin.java] and later deployed as .jar files in your cluster. 
```

Comment:
`by declaring them in [UdfPlugin.java]` -> `by declaring them in the [UdfPlugin.java] file and then deploying them as a .jar file`